fendermint/vm/message/src/ipc.rs

```diff
@@ -0,0 +1,169 @@
+
+    /// A bottom-up checkpoint proposed "for execution" by the parent subnet validators, provided that the majority of them
+    /// have the data available to them, already resolved.
+    ///
+    /// To prove that the data is available, we can either use the ABCI++ "process proposal" mechanism,
+    /// or we can gossip votes using the _IPLD Resolver_ and attach them as a quorum certificate.
+    BottomUpExec(CertifiedMessage<BottomUpCheckpoint>),
+
+    // TODO
+    TopDown,
+}
+
+/// A message relayed by a user on the current subnet.
+///
+/// The relayer pays for the inclusion of the message in the ledger,
+/// but not necessarily for the execution of its contents.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct RelayedMessage<T> {
+    /// The relayed message.
+    pub message: T,
+    /// The address (public key) of the relayer in the current subnet.
+    pub relayer: Address,
+    /// The nonce of the relayer in the current subnet.
+    pub sequence: u64,
+}
+
+/// Relayed messages are signed by the relayer, so we can rightfully charge them message inclusion costs.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct SignedRelayedMessage<T> {
+    /// The relayed message with the relayer identity.
+    pub message: RelayedMessage<T>,
+    /// The signature of the relayer, for cost and reward attribution.
+    pub signature: Signature,
+}
+
+/// A message with a quorum certificate from a group of validators.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CertifiedMessage<T> {
+    /// The message the validators signed.
+    pub message: T,
+    /// The quorum certificate.
+    pub certificate: MultiSig,
+}
+
+/// A quorum certificate consisting of a simple multi-sig.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct MultiSig {
+    pub signatures: Vec<ValidatorSignature>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ValidatorSignature {
+    pub validator: Address,
+    pub signature: Signature,
+}
+
+/// A periodic bottom-up checkpoints contains the source subnet ID (to protect against replay attacks),
+/// a block height (for sequencing), any potential handover to the next validator set, and a pointer
+/// to the messages that need to be resolved and executed by the parent validators.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct BottomUpCheckpoint {
+    /// Which subnet is the checkpoint coming from.
+    pub subnet_id: SubnetID,
+    /// Block height of this checkpoint.
+    pub height: ChainEpoch,
+    /// Which validator set is going to sign the *next* checkpoint.
+    /// The parent subnet already expects the last validator set to sign this one.
+    pub next_validator_set_id: u64,
+    /// Pointer at all the bottom-up messages included in this checkpoint.
+    pub bottom_up_messages: Cid, // TODO: Use TCid
```

Comment:
We may need to include the list of messages here and not just the cid for L2->root interactions as described above

Reply:
This data structure doesn't have to be the be-all, end-all data structure. It is for bottom-up messages _arriving_ into Fendermint, and that is all we care about. How it is managed with Lotus is the concern of those relayers which send this to Lotus. For example they can do the equivalent of this:

```rust
pub struct RelayedCheckpoint {
  pub checkpoint: CertifiedMessage<BottomUpCheckpoint>,
  /// The CID of these messages have to match the CID in the checkpoint.
  pub bottom_up_messages: Vec<CrossMsg>
}
```

And then part of validation is to check that. 

I know it sucks that you have a single set of Solidity actors and two different ideas for relaying the stuff :disappointed: